pkg/cache/v2/simple_delta.go

```diff
@@ -0,0 +1,297 @@
+					r := Resources{
+						Version: version,
+					}
+					for key, value := range diff {
+						for rKey := range info.deltaState[t].Items {
+							// Handle the case when a new item could be added to the state and also if we need to overwrite a previous resource
+							if key == rKey {
+								info.deltaState[t].Items[key] = value
+							} else if _, found := info.deltaState[t].Items[key]; !found {
+								info.deltaState[t].Items[key] = value
+							}
+						}
+					}
+					r.Items = info.deltaState[t].Items
+
+					info.deltaState[t] = r
+				}
+
+				if len(unsubscribed) > 0 {
+					// we need to remove the previously subscribed resources from the state so we no longer send updates
+					if cache.log != nil {
+						cache.log.Debugf("node: %s, recieved items to unsubscribe from: %v", node, unsubscribed)
+					}
+
+					newState := cache.unsubscribe(unsubscribed, info.deltaState)
+					for typ := range info.deltaState {
+						info.deltaState[typ] = newState[typ]
+					}
+				}
+
+				if cache.log != nil {
+					// We only want to show the specific resources we're sending back from the diff
+					cache.log.Debugf("delta respond open watch ID:%d Resources:%+v with new version %q", id, diff, version)
+				}
+
+				// Respond to our delta stream with the subcribed resources
+				cache.respondDelta(
+					watch.Request,
+					watch.Response,
+					diff, // We want to only send the updated resources here
+					unsubscribed,
+					version,
+				)
+
+				// discard the old watch
+				delete(info.deltaWatches, id)
+			}
+		}
+		info.mu.Unlock()
+	}
+
+	return nil
+}
+
+// difference returns the elements in `a` that aren't in `b`.
+// TODO: SLOW this will need to be revisited
+func (cache *snapshotCache) checkState(resources, deltaState map[string]types.Resource) map[string]types.Resource {
+	mb := make(map[string]types.Resource, len(deltaState))
+	diff := make(map[string]types.Resource, 0)
+
+	for key, value := range deltaState {
+		mb[key] = value
+	}
+
+	// Check our diff map to see what has changed
+	// Even is an underlying resource has changed we need to update the diff
+	for key, value := range resources {
+		if _, found := mb[key]; !found {
+			cache.log.Debugf("found new key: %s", key)
+			diff[key] = value
+		} else if resource, found := mb[key]; found && (resource != value) {
+			cache.log.Debugf("found updated resource from existing key %s, modifying resource map", key)
+			diff[key] = value
+		}
+	}
+
+	return diff
+}
+
+// Unscrubscribe will remove resources from the tracked list in the management server when received from a client
+func (cache *snapshotCache) unsubscribe(resources []string, deltaState map[string]Resources) map[string]Resources {
+	// here we need to search and remove from the current subscribed list in the snapshot
+	for _, items := range deltaState {
+		for i := 0; i < len(resources); i++ {
+			delete(items.Items, resources[i])
+		}
+	}
+
+	return deltaState
```

Comment:
Not sure why you need to return `deltaState` here as you have mutated the input `deltaState` so the caller can access the updated state anyway.

Reply:
Yup thank you, removed that